tools/test_history.py

```diff
@@ -0,0 +1,160 @@
+    job: str
+) -> Any:
+    objs = list(bucket.objects.filter(Prefix=f"test_time/{sha}/{job}"))
+    if len(objs) == 0:
+        return {}
+    return json.loads(bz2.decompress(objs[0].get()['Body'].read()))
+
+
+def case_status(case: Dict[str, Any]) -> Optional[str]:
+    for k in {'errored', 'failed', 'skipped'}:
+        if case[k]:
+            return k
+    return None
+
+
+def make_column(
+    *,
+    bucket: Any,
+    sha: str,
+    job: str,
+    suite_name: str,
+    test_name: str,
+    digits: int,
+) -> str:
+    decimals = 3
+    num_length = digits + 1 + decimals
+
+    data = get_ossci_json(bucket=bucket, sha=sha, job=job)
+    suite = data.get('suites', {}).get(suite_name)
+    if suite:
+        testcase_times = {
+            case['name']: case
+            for case in suite['cases']
+        }
+        case = testcase_times.get(test_name)
+        if case:
+            status = case_status(case)
+            if status:
+                return f'{status.rjust(num_length)} '
+            else:
+                return f'{case["seconds"]:{num_length}.{decimals}f}s'
+    return ' ' * (num_length + 1)
+
+
+def display_history(
+    *,
+    bucket: Any,
+    commits: List[Tuple[str, datetime]],
+    jobs: List[str],
+    suite_name: str,
+    test_name: str,
+    delta: int,
+    digits: int,
+) -> None:
+    prev_time = datetime.now()
+    for sha, time in commits:
+        if (prev_time - time).total_seconds() < delta * 3600:
+            continue
+        prev_time = time
+        columns = [
+            make_column(
+                bucket=bucket,
+                sha=sha,
+                job=job,
+                suite_name=suite_name,
+                test_name=test_name,
+                digits=digits,
+            )
+            for job in jobs
+        ]
+        print(f"{time} {sha} {' '.join(columns)}".rstrip())
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        __file__,
+        description='Display the history of a test.',
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+    )
+    parser.add_argument(
+        '--pytorch',
+        help='path to local PyTorch clone',
+        default='.',
+    )
+    parser.add_argument(
+        '--ref',
+        help='starting point (most recent Git ref) to display history for',
+        default='master',
+    )
+    parser.add_argument(
+        '--delta',
+        type=int,
+        help='minimum number of hours between rows',
+        default=12,
+    )
+    parser.add_argument(
+        '--digits',
+        type=int,
+        help='number of digits to display before the decimal point',
+        default=4,
+    )
+    parser.add_argument(
+        'suite',
+        help='name of the suite containing the test',
+    )
+    parser.add_argument(
+        'test',
+        help='name of the test',
+    )
+    parser.add_argument(
+        'job',
+        nargs='*',
+        help='names of jobs to display columns for, in order',
+    )
```

Comment:
it would probably be easier to not ask for job and just display all of them in single rows?

Reply:
separate rows? I think that'd be much much harder to read

Reply:
It might be hard for ppl to read multiple column anyway since i need to remember exactly what order i put in.
IMO, printing everything out in a single row is like a standard database table format as each row contains all info. 

But I am open to the current solution too.

Reply:
I think ideally this would be presented on a webpage like Ed's HUD, so you could just hover over a column to see which job it corresponds to; but if we're trying to make it readable as a CLI script, I prefer the current solution. Could you give an example of a use case that is better served by what you're suggesting? Maybe I'm not quite seeing where you're coming from.

Reply:
hmm. I am a bit confused. is this intended for a debugging tool on CLI or display on HUD? and if latter, how would that work? do we run this on every single test and list them out in a dashboard (that would be a huge list consider we have more than 20K+)

My point with the CLI is that with the current snippet, I need to remember what order of the jobs I put in, or scroll the screen up to check which number corresponds to which job. it might be easier if they are printed in the same line.

Reply:
> is this intended for a debugging tool on CLI or display on HUD?

Sorry, my previous reply wasn't very clear. What I meant was that if this were on a webpage like the HUD then we could do nice things like display the name of the job on hover, but since this is just a CLI script, we can't do that here.

> My point with the CLI is that with the current snippet, I need to remember what order of the jobs I put in, or scroll the screen up to check which number corresponds to which job. it might be easier if they are printed in the same line.

I agree that that's suboptimal, but my point is that it would be worse to display them on separate lines. For instance, if we have two jobs `job1` and `job2` where `job2` runs on a much more powerful machine, we would get this annoying interleaving in the output:
```
time commit job1 123.45s
time commit job2   2.34s
time commit job1 123.45s
time commit job2   2.34s
time commit job1 123.45s
time commit job2   2.34s
time commit job1 123.45s
time commit job2   2.34s
```
I find that harder to read than what the script currently does:
```
time commit 123.45s   2.34s
time commit 123.45s   2.34s
time commit 123.45s   2.34s
time commit 123.45s   2.34s
```

Reply:
> > is this intended for a debugging tool on CLI or display on HUD?
> 
> Sorry, my previous reply wasn't very clear. What I meant was that if this were on a webpage like the HUD then we could do nice things like display the name of the job on hover, but since this is just a CLI script, we can't do that here.
> 
> > My point with the CLI is that with the current snippet, I need to remember what order of the jobs I put in, or scroll the screen up to check which number corresponds to which job. it might be easier if they are printed in the same line.
> 
> I agree that that's suboptimal, but my point is that it would be worse to display them on separate lines. For instance, if we have two jobs `job1` and `job2` where `job2` runs on a much more powerful machine, we would get this annoying interleaving in the output:
> 
> ```
> time commit job1 123.45s
> time commit job2   2.34s
> time commit job1 123.45s
> time commit job2   2.34s
> time commit job1 123.45s
> time commit job2   2.34s
> time commit job1 123.45s
> time commit job2   2.34s
> ```
The interleaving nature can be mitigated by `| grep `. e.g,. I can grep by job1 and it will be nicely 
```
time commit1 job1  123.45s
time commit2 job1 123.45s
```
or by commits I can see all relevant jobs. 
```
time commit1 job1  123.45s
time commit1 job2 2.345s
```
> 
> ```
> time commit 123.45s   2.34s
> time commit 123.45s   2.34s
> time commit 123.45s   2.34s
> time commit 123.45s   2.34s
> ```
i guess my main concern is I can't further process this output with CLI tricks

Reply:
Right, but what about the use case where someone actually wants to see multiple jobs at once?